### PR TITLE
Ray PDB support

### PR DESCRIPTION
**Pull request by pcmoritz**
## Why are these changes needed?

This PR adds PDB support for Ray. I will follow up with more detailed documentation but here is the gist of how you can use it: Say you have a Ray program like
```python
import ray

ray.init()

@ray.remote
def f():
    i = 1
    tasks = [g.remote(j) for j in range(5)]
    ray.util.pdb.set_trace()

@ray.remote
def g(j):
    h = 1
    if j == 3:
        ray.util.pdb.set_trace()

ray.get([f.remote()])
```

Execute it with `python <name>.py` and the program will drop into the debugger. You can attach to the debugger with `ray debug` in another terminal. It will print
```
2020-11-01 15:18:04,598	INFO worker.py:674 -- Connecting to existing Ray cluster at address: 192.168.1.105:6379
Active breakpoints:
0: ray::g() | test.py:15

1: ray::f() | test.py:9

Enter breakpoing index or press enter to refresh:
```

You can now press `0 <ENTER>` to enter the debugger in task `g` and `1 <ENTER>` to enter the debugger in task `f`.

You can also do post-mortem debugging. Take e.g. this Ray serve script:

```python
import ray
from ray import serve
import pickle
import requests
from sklearn.datasets import load_iris
from sklearn.ensemble import GradientBoostingClassifier
import time

# Train model
iris_dataset = load_iris()
model = GradientBoostingClassifier()
model.fit(iris_dataset["data"], iris_dataset["target"])

# Define Ray Serve model,
class BoostingModel:
    def __init__(self):
        self.model = model
        self.label_list = iris_dataset["target_names"].tolist()

    def __call__(self, flask_request):
        payload = flask_request.json["vector"]
        print("Worker: received flask request with data", payload)

        prediction = self.model.predict([payload])[0]
        human_name = self.label_list[prediction]
        return {"result": human_name}


# Deploy model
client = serve.start()
client.create_backend("iris:v1", BoostingModel)
client.create_endpoint("iris_classifier", backend="iris:v1", route="/iris")

time.sleep(3600.0)
```

and execute it with `RAY_PDB=1 python <name>.py`. Send an invalid request
```
python -c 'import requests; response = requests.get("http://localhost:8000/iris", json={"vector": [1.2, 1.0, 1.1, "a"]})'
```

You can now again debug with `ray debug`, without having to have set a breakpoint!

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/rpdb.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
+# Some code in this file is from
```
Comment (ericl) [inline python/ray/rpdb.py:1]: Can we put this in `ray/util`?

File: python/ray/scripts/scripts.py [hunk 2]
```diff
@@ -150,6 +150,44 @@ def dashboard(cluster_config_file, cluster_name, port, remote_port):
...
+            "RAY_PDB_")
+        print("Active breakpoints:")
+        for i, active_session in enumerate(active_sessions):
+            data = json.loads(
+                ray.experimental.internal_kv._internal_kv_get(active_session))
+            print(
+                str(i) + ": " + data["proctitle"] + " | " + data["filename"] +
+                ":" + str(data["lineno"]))
+            print(data["traceback"])
+        inp = input("Enter breakpoing index or press enter to refresh: ")
```
Comment (ericl) [inline python/ray/scripts/scripts.py:177]:
```suggestion
        inp = input("Enter breakpoint index or press enter to refresh: ")
```

File: python/ray/tests/test_ray_debugger.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import json
```
Comment (ericl) [inline python/ray/tests/test_ray_debugger.py:1]: You'll need to add this to the BUILD file for travis to run this

File: python/ray/rpdb.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+import traceback
+
+from ray.experimental.internal_kv import _internal_kv_del, _internal_kv_put
+
+PY3 = sys.version_info[0] == 3
+log = logging.getLogger(__name__)
+
+
+def cry(message, stderr=sys.__stderr__):
+    log.critical(message)
```
Comment (ericl) [inline python/ray/rpdb.py:25]: Why critical?

Comment (ericl) [reply python/ray/rpdb.py:25]: Actually, we shouldn't need to log since the print() below shows up already.

File: python/ray/rpdb.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        self._quiet = quiet
+        self._patch_stdstreams = patch_stdstreams
+        self._listen_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self._listen_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR,
+                                       True)
+        self._listen_socket.bind((host, port))
+
+    def listen(self):
+        if not self._quiet:
+            cry("RemotePdb session open at %s:%s, waiting for connection ..." %
```
Comment (ericl) [inline python/ray/rpdb.py:84]: Can you also include "use `ray debug`" to connect? It's not intuitive you have to do that.

**Comment by pcmoritz** [comment]
Thanks a lot for the comments @ericl, they should now all be addressed :)

**Review by ericl — APPROVED** [review]

**Pull request merged**